python/lsst/daf/butler/core/datastoreRecordData.py

```diff
@@ -82,15 +120,29 @@ class DatastoreRecordData:
     datastore.
     """
 
-    refs: List["DatasetIdRef"] = dataclasses.field(default_factory=list)
-    """List of DatasetRefs known to this datastore.
+    dataset_ids: Set[DatasetId] = dataclasses.field(default_factory=set)
+    """List of DatasetIds known to this datastore.
     """
 
-    records: Dict[str, List[StoredDatastoreItemInfo]] = dataclasses.field(
-        default_factory=lambda: defaultdict(list)
+    records: Dict[DatasetId, Dict[str, List[StoredDatastoreItemInfo]]] = dataclasses.field(
```

Comment:
I think it'd be better to annotate this as:
```
defaultdict[DatasetId, defaultdict[str, list[StoredDatastoreItemInfo]]]
```
i.e. explicitly point out that these are `defautldict` objects, since (despite the formal inheritance relationship) the differing behavior of `__getitem__` can be confusing when you're expecting a vanilla `dict`.  Even more importantly, (as I think another comment here demonstrates) if someone did put a vanilla `dict` here and other code was expecting to use `defaultdict` for the usual missing-insertion pattern it provides, that code would break, and we want MyPy to catch that.